server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -102,14 +122,141 @@ public void writeFileChunk(
         int totalTranslogOps,
         ActionListener<Void> listener
     ) {
-        // TODO
+        try {
+            multiFileWriter.writeFileChunk(metadata, position, content, lastChunk);
+            listener.onResponse(null);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
     }
 
     /**
      * Start the Replication event.
      * @param listener {@link ActionListener} listener.
      */
     public void startReplication(ActionListener<Void> listener) {
-        // TODO
+        state.setStage(SegmentReplicationState.Stage.REPLICATING);
+        final StepListener<CheckpointInfoResponse> checkpointInfoListener = new StepListener<>();
+        final StepListener<GetSegmentFilesResponse> getFilesListener = new StepListener<>();
+        final StepListener<Void> finalizeListener = new StepListener<>();
+
+        // Get list of files to copy from this checkpoint.
+        source.getCheckpointMetadata(getId(), checkpoint, checkpointInfoListener);
+
+        checkpointInfoListener.whenComplete(checkpointInfo -> getFiles(checkpointInfo, getFilesListener), listener::onFailure);
+        getFilesListener.whenComplete(
+            response -> finalizeReplication(checkpointInfoListener.result(), finalizeListener),
+            listener::onFailure
+        );
+        finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
+    }
+
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
```

Comment:
My main concern with these private methods are they probably aren't getting unit tested

Reply:
While the private methods aren't being tested directly, the implementation details are being tested. For example - this part of the code checks if the files fetched are what we expected - https://github.com/opensearch-project/OpenSearch/blob/bd9bf6797719cadd66400626328f6b5c383c5d79/server/src/test/java/org/opensearch/indices/replication/SegmentReplicationTargetTests.java#L121-L123